tests/connect_redirect/connect_redirect_tests.cpp

```diff
@@ -0,0 +1,422 @@
+    _In_ sockaddr_storage& proxy,
+    uint16_t destination_port,
+    uint16_t proxy_port,
+    bool dual_stack)
+{
+    bool add_policy = true;
+    // Update policy in the map to redirect the connection to the proxy.
+    _update_policy_map(
+        object, destination, proxy, destination_port, proxy_port, _globals.protocol, dual_stack, add_policy);
+
+    // Try to send and receive message to "destination". It should succeed.
+    sender_socket->send_message_to_remote_host(CLIENT_MESSAGE, destination, _globals.destination_port);
+    sender_socket->complete_async_send(1000, expected_result_t::SUCCESS);
+
+    sender_socket->post_async_receive();
+    sender_socket->complete_async_receive(2000, false);
+
+    uint32_t bytes_received = 0;
+    char* received_message = nullptr;
+    sender_socket->get_received_message(bytes_received, received_message);
+
+    std::string expected_response = SERVER_MESSAGE + std::to_string(proxy_port);
+    REQUIRE(strlen(received_message) == strlen(expected_response.c_str()));
+    REQUIRE(memcmp(received_message, expected_response.c_str(), strlen(received_message)) == 0);
+
+    // Remove entry from policy map.
+    add_policy = false;
+    _update_policy_map(
+        object, destination, proxy, destination_port, proxy_port, _globals.protocol, dual_stack, add_policy);
+}
+
+void
+authorize_test(
+    _In_ const struct bpf_object* object,
+    _In_ client_socket_t* sender_socket,
+    _Inout_ sockaddr_storage& destination,
+    bool dual_stack)
+{
+    // Default behavior of the eBPF program is to block the connection.
+
+    // Send should fail as the connection is blocked.
+    sender_socket->send_message_to_remote_host(CLIENT_MESSAGE, destination, _globals.destination_port);
+    sender_socket->complete_async_send(1000, expected_result_t::FAILURE);
+
+    // Receive should timeout as connection is blocked.
+    sender_socket->post_async_receive(true);
+    sender_socket->complete_async_receive(1000, true);
+
+    // Now update the policy map to allow the connection and test again.
+    connect_redirect_test(
+        object,
+        sender_socket,
+        destination,
+        destination,
+        _globals.destination_port,
+        _globals.destination_port,
+        dual_stack);
+}
+
+void
+get_client_socket(bool dual_stack, _Inout_ client_socket_t** sender_socket)
+{
+    client_socket_t* old_socket = *sender_socket;
+    client_socket_t* new_socket = nullptr;
+    socket_family_t family = dual_stack
+                                 ? socket_family_t::Dual
+                                 : ((_globals.family == AF_INET) ? socket_family_t::IPv4 : socket_family_t::IPv6);
+    if (_globals.protocol == IPPROTO_TCP) {
+        new_socket = (client_socket_t*)new stream_client_socket_t(SOCK_STREAM, IPPROTO_TCP, 0, family);
+    } else {
+        new_socket = (client_socket_t*)new datagram_client_socket_t(SOCK_DGRAM, IPPROTO_UDP, 0, family);
+    }
+
+    *sender_socket = new_socket;
+    if (old_socket) {
```

Comment:
Ok. but note that "delete" is safe to pass a nullptr to so can reduce code paths by removing this check.